File: 12/main.rb
def read line
  f = line.split(' ')
  a = [f[0], f[0], f[0], f[0], f[0]].join "?"
  b = [f[1], f[1], f[1], f[1], f[1]].join ","
  return "#{a} #{b}"
end

def is_valid? arrangement, target_arrangement
  hash_counts = arrangement.split(/\.+/).select{|s| s.length > 0}.collect{|s| s.length}

  if hash_counts.length != target_arrangement.length then
    return false
  end

  hash_counts.each_index do |i|
    hash_count = hash_counts[i]
    target_count = target_arrangement[i]

    if hash_count != target_count then
      return false
    end
  end

  return true
end

def evaluate inlet
  fields = inlet.split(' ')
  raw_springs = fields[0]
  target_arrangement = fields[1].split(',').collect{|x| x.to_i}
  possible_arrangements = 2**(raw_springs.split('').select{|x|x=='?'}.count)
  outlet = 0

  for i in 0...possible_arrangements do
    j = 0
    arrangement = ""
    raw_springs.each_char do |c|
      if c == '?' then
        d  = (((i >> j) & 0x1) == 0x1)? '#' : '.'
        arrangement << d
        j += 1
      else
        arrangement << c
      end
    end
    if is_valid? arrangement, target_arrangement
      outlet += 1
    end
  end
  
  return outlet
end

def main
  outlet = 0
  line = gets
  until line.nil? do
    inlet = read line.chomp
    outlet += evaluate inlet
    line = gets
  end
  puts outlet
end

if __FILE__ == $0 then
  main
end

File: 12/main.rs
use std::io::{self, BufRead};

fn read_arrangement(inlet: String) -> Vec<i32> {
    let mut outlet: Vec<i32> = Vec::new();
    let fields: Vec<&str> = inlet.split(",").collect();

    for field in fields {
        outlet.push(field.parse().unwrap());
    }

    return outlet;
}

fn evaluate(inlet: String) -> i32 {
    let fields: Vec<&str> = inlet.split(" ").collect();
    let raw_springs: String = fields[0].to_string();
    let target_arrangement: Vec<i32> = read_arrangement(fields[1].to_string());
    let mut possible_arrangements: i32 = 0;

    // count possible arrangements
    for c in raw_springs.chars() {
        if c == '?' {
            possible_arrangements += 1;
        }
    }
    
    // TODO generate all possible arrangements, counting the valid ones

    return 0;
}

fn main() {
    let stdin = io::stdin();
    let mut inlet: String;
    let mut outlet: i32 = 0;

	for line in stdin.lock().lines() {
        inlet = line.unwrap();
        if inlet.len() > 0 {
            outlet += evaluate(inlet.clone());
        }
	}

    println!("{}", outlet);
}

